lib/stages/Writer.cpp

```diff
@@ -114,76 +114,153 @@ void visWriter::main_thread() {
         }
 
         // Get a view of the current frame
-        auto frame = VisFrameView(in_buf, frame_id);
-
-        acqs_lock.lock();
-        // Check the dataset ID hasn't changed
-        if (acqs.count(frame.dataset_id) == 0) {
-            init_acq(frame.dataset_id);
+        if (!strcmp(in_buf->buffer_type, "vis")) {
+            auto frame = VisFrameView(in_buf, frame_id);
+            write_vis_data(frame, write_time_metric, acqs_lock);
+        } else if (!strcmp(in_buf->buffer_type, "hfb")) {
+            auto frame = HFBFrameView(in_buf, frame_id);
+            write_hfb_data(frame, write_time_metric, acqs_lock);
         }
```

Comment:
I really don't like this (including the associated `write_X_data` routines). It's not clear this is any real improvement on just having separate stages for this.

The case I would like to think of is someone wanting to implement a new data/buffer type that wants to use all of this functionality, but isn't able to modify the source of this function, i.e. how can you do that in a purely pluggable format.

I can think of two vague directions that might work. One is moving as much functionality into a generic base that's templated on the frame view type, and then probably aliasing the templated types to different names to register separate processes.

Another is to move the format specific functionality into a separate class, which uses an abstract factory, and then use the buffer type to dynamically create the correct help class for writing the data.

There may be other options.

Reply:
Yeah this is the bit I wasn't sure how to handle correctly. So you would have a generic base class that `Writer.cpp` creates? Are there any other examples in `kotekan` that do this?